catalyst_voices_packages/catalyst_cardano_serialization/lib/src/witness.dart

```diff
@@ -1,35 +1,150 @@
-import 'package:catalyst_cardano_serialization/src/signature.dart';
-import 'package:catalyst_cardano_serialization/src/types.dart';
+import 'package:catalyst_cardano_serialization/catalyst_cardano_serialization.dart';
 import 'package:cbor/cbor.dart';
 import 'package:equatable/equatable.dart';
 
+/// Witnesses that allow for various types of transactions on the Cardano
+/// blockchain.
+enum WitnessType {
+  /// A witness that contains a verification key.
+  vkeyWitness(0),
+
+  /// A native script witness.
+  nativeScript(1),
+
+  /// A bootstrap witness.
+  bootstrapWitness(2),
+
+  /// A Plutus v1 script witness.
+  plutusV1Script(3),
+
+  /// A Plutus data witness.
+  plutusData(4),
+
+  /// A redeemer witness.
+  redeemers(5),
+
+  /// A Plutus v2 script witness.
+  plutusV2Script(6),
+
+  /// A Plutus v3 script witness.
+  plutusV3Script(7);
+
+  /// The integer value associated with the witness type.
+  final int value;
+
+  /// Constructs a new `WitnessType` with the given integer value.
+  const WitnessType(this.value);
+}
+
 /// A set of witnesses that sign the transaction.
 final class TransactionWitnessSet extends Equatable implements CborEncodable {
   /// The witnesses that sign the transaction.
   final Set<VkeyWitness> vkeyWitnesses;
 
+  /// The native scripts.
+  final Set<NativeScript> nativeScripts;
+  // TODO(ilap): implement bootstrap witnesses.
+  // final Set<BootstrapWitness> bootstrapWitnesses;
+  /// The Plutus V1 scripts.
+  final Set<PlutusV1Script> plutusV1Scripts;
+  // TODO(ilap): implement plutus data.
+  // final Set<PlutusData> plutusDatas;
```

Comment:
Consider to remove commented out code for the PR to allow more transparent review, you can keep those changes as a stash on your branch if this is still work in progress